cartridge/hotreload.lua

```diff
@@ -0,0 +1,194 @@
+
+--- Snapshot names of all running fibers.
+--
+-- @function snap_fibers
+-- @treturn {string=true,...} Set of fiber names
+local function snap_fibers()
+    local ret = {}
+
+    for _, f in pairs(fiber.info()) do
+        ret[f.name] = true
+    end
+
+    return ret
+end
+
+--- Get difference between two snapshots.
+-- @function diff
+-- @tparam {string=true,...} snap1
+-- @tparam {string=true,...} snap2
+-- @treturn {string,...} Keys from snap2 that are not in snap1, sorted
+local function diff(t1, t2)
+    local ret = {}
+    for k, _ in pairs(t2) do
+        if not t1[k] then
+            table.insert(ret, k)
+        end
+    end
+
+    table.sort(ret)
+    return ret
+end
+
+--- Avoid removing globals with given names.
+--
+-- @function whitelist_globals
+-- @tparam {string,...} names
+local function whitelist_globals(tbl)
+    if not vars.globals then
+        return
+    end
+
+    for _, v in ipairs(tbl) do
+        log.debug('Avoid removing global %q on hot-reload', v)
+        vars.globals[v] = true
+    end
+end
+
+--- Avoid cancelling fibers with given names.
+--
+-- Names 'lua', 'main' and '* (net.box)' are ignored and can't be
+-- whitelisted.
+--
+-- @function whitelist_fibers
+-- @tparam {string,...} names
+local function whitelist_fibers(tbl)
+    if not vars.fibers then
+        return
+    end
+
+    for _, v in ipairs(tbl) do
+        if v == 'lua'
+        or v == 'main'
+        or v:endswith('(net.box)')
+        then
+            log.debug('Refusing to whitelist fiber %q', v)
+        else
+            log.debug('Avoid cancelling fiber %q on hot-reload', v)
+            vars.fibers[v] = true
+        end
+    end
+end
+
+--- Check that state was saved.
+--
+-- @function state_saved
+-- @treturn boolean true / false
+local function state_saved()
+    return vars.packages
+        and vars.globals
+        and vars.fibers
+        and vars.routes_count
+end
+
+--- Save initial state.
+--
+-- This includes `package.loaded`, `_G`, fibers, http routes.
+--
+-- @function save_state
+local function save_state()
+    vars.packages = table.copy(package.loaded)
+    for k, _ in pairs(vars.packages) do
+        vars.packages[k] = true
+    end
+
+    vars.globals = setmetatable(table.copy(_G), nil)
+    for k, _ in pairs(vars.globals) do
+        vars.globals[k] = true
+    end
+
+    vars.fibers = {}
+    whitelist_fibers(diff(vars.fibers, snap_fibers()))
+
+    local httpd = service_registry.get('httpd')
+    if httpd ~= nil then
+        vars.routes_count = #httpd.routes
+    else
+        vars.routes_count = 0
+    end
+
+    assert(state_saved())
+end
+
+--- Restore initial state.
+--
+-- Unload packages, cancel fibers, remove globals and http routes.
+--
+-- @function load_state
+local function load_state()
+    assert(state_saved(), "Hot-reload state wasn't saved")
+
+    for k, _ in pairs(package.loaded) do
+        if not vars.packages[k] then
+            -- log.info('Unloading package %q', k)
```

Comment:
Please drop if it's redundant. The same for https://github.com/tarantool/cartridge/pull/1105/files#diff-589dc35678beb0f0e0367553c5baab56f0ba15cc332d830bc5464f670dfad33bR140